Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMSessionEncKey.cpp

```diff
@@ -0,0 +1,110 @@
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebKitOpenCDMSessionEncKey.h"
+
+#if ENABLE(LEGACY_ENCRYPTED_MEDIA) && USE(OCDM)
+
+#include "CDM.h"
+#include "CDMSession.h"
+#include <gst/gst.h>
+#include "WebKitMediaKeyError.h"
+
+GST_DEBUG_CATEGORY_EXTERN(webkit_media_opencdm_decrypt_debug_category);
+#define GST_CAT_DEFAULT webkit_media_opencdm_decrypt_debug_category
+
+namespace WebCore {
+
+WebKitOpenCDMSessionEncKey::WebKitOpenCDMSessionEncKey(CDMSessionClient* client, OpenCdm* openCdm)
+    : m_client(client)
+    , m_openCdmSession(openCdm)
+    , m_destinationUrlLength(0)
+{
+}
+
+RefPtr<Uint8Array> WebKitOpenCDMSessionEncKey::generateKeyRequest(const String& mimeType, Uint8Array* initData, String& destinationUrl, unsigned short& errorCode, uint32_t&)
+{
+    int returnValue = 0;
+
+    unsigned char initDataValue[initData->length()];
+
+    memcpy(initDataValue, initData->data(), initData->length());
+
+    string sessionId;
+    m_openCdmSession->CreateSession(mimeType.utf8().data(), initDataValue,
+        initData->length(), sessionId);
+    if (!sessionId.size()) {
+        GST_WARNING("SessionId is empty\n");
+        return nullptr;
+    }
+    m_sessionId = String::fromUTF8(sessionId.c_str());
+
+    unsigned char temporaryUrl[1024] = "\0";
+    string message;
+    int messageLength = 0;
+    returnValue = m_openCdmSession->GetKeyMessage(message,
+        &messageLength, temporaryUrl, &m_destinationUrlLength);
+    if (returnValue || !messageLength || !m_destinationUrlLength) {
+        errorCode = WebKitMediaKeyError::MEDIA_KEYERR_UNKNOWN;
+        return nullptr;
+    }
+    else {
+        destinationUrl = String::fromUTF8(temporaryUrl);
+    }
+    return Uint8Array::create(reinterpret_cast<unsigned char*>(const_cast<char*>(message.c_str())), messageLength);
+}
+
+void WebKitOpenCDMSessionEncKey::releaseKeys()
+{
+}
+
+bool WebKitOpenCDMSessionEncKey::update(Uint8Array* key, RefPtr<Uint8Array>& nextMessage, unsigned short& errorCode, uint32_t& sysytemCode)
+{
+    UNUSED_PARAM(sysytemCode);
+
+    GST_MEMDUMP("License :", key->data(), key->length());
+
+    int returnValue = 0;
+    std::string responseMessage;
+    returnValue = m_openCdmSession->Update(key->data(), key->length(), responseMessage);
+    if (returnValue) {
+        responseMessage = "UpdateStatus: " + responseMessage;
+        String prettyResponseMessage = responseMessage.c_str();
```

Comment:
I meant to write it otherwise. I tried it and it does not bring any advantage.

Please remove prettyResponseMessage and reinterpret_cast responseMessage.c_str() when calling create.